pkg/controller/elasticsearch/pdb/reconcile.go

```diff
@@ -7,107 +7,165 @@ package pdb
+	if err := controllerutil.SetControllerReference(&es, &expected, scheme); err != nil {
+		return nil, err
+	}
+
+	if template.Spec.Selector != nil || template.Spec.MaxUnavailable != nil || template.Spec.MinAvailable != nil {
+		// use the user-defined spec
+		expected.Spec = template.Spec
+		return &expected, nil
+	}
+
+	// set our default spec
+	var err error
+	if expected.Spec, err = buildPDBSpec(es, statefulSets); err != nil {
+		return nil, err
+	}
+
+	return &expected, nil
+}
+
+// buildPDBSpec returns a PDBSpec computed from the current StatefulSets,
+// considering the cluster health and topology.
+func buildPDBSpec(es v1alpha1.Elasticsearch, statefulSets sset.StatefulSetList) (v1beta1.PodDisruptionBudgetSpec, error) {
+	// compute MinAvailable based on the maximum number of Pods we're supposed to have
+	nodeCount := statefulSets.ExpectedPodCount()
+	// maybe allow some Pods to be disrupted
+	minAvailable := nodeCount - allowedDisruptions(es, statefulSets)
+
+	minAvailableIntStr := intstr.IntOrString{Type: intstr.Int, IntVal: minAvailable}
+
+	return v1beta1.PodDisruptionBudgetSpec{
+		// match all pods for this cluster
+		Selector: &metav1.LabelSelector{
 			MatchLabels: map[string]string{
 				label.ClusterNameLabelName: es.Name,
 			},
-		}
-	}
-
-	var reconciled v1beta1.PodDisruptionBudget
-	return reconciler.ReconcileResource(reconciler.Params{
-		Client:     c,
-		Scheme:     scheme,
-		Owner:      &es,
-		Expected:   &expected,
-		Reconciled: &reconciled,
-		NeedsUpdate: func() bool {
-			for k, v := range expected.Labels {
-				if rv, ok := reconciled.Labels[k]; !ok || rv != v {
-					return true
-				}
-			}
-
-			return !reflect.DeepEqual(expected.Spec, reconciled.Spec)
 		},
-		UpdateReconciled: func() {
-			for k, v := range expected.Labels {
-				reconciled.Labels[k] = v
-			}
-			reconciled.Spec = expected.Spec
-		},
-	})
+		MinAvailable: &minAvailableIntStr,
+		// MaxUnavailable can only be used if the selector matches a builtin controller selector
+		// (eg. Deployments, StatefulSets, etc.). We cannot use it with our own cluster-name selector.
+		MaxUnavailable: nil,
+	}, nil
+}
+
+// allowedDisruptions returns the number of Pods that we allow to be disrupted while keeping the cluster healthy.
+func allowedDisruptions(es v1alpha1.Elasticsearch, actualSsets sset.StatefulSetList) int32 {
+	if es.Status.Health != v1alpha1.ElasticsearchGreenHealth {
+		// A non-green cluster may become red if we disrupt one node, don't allow it.
+		// The health information we're using here may be out-of-date, that's best effort.
+		return 0
+	}
+	if actualSsets.ExpectedMasterNodesCount() == 1 {
+		// There's a risk the single master of the cluster gets removed, don't allow it.
+		return 0
+	}
+	if actualSsets.ExpectedDataNodesCount() == 1 {
+		// There's a risk the single data node of the cluster gets removed, don't allow it.
+		return 0
+	}
+	// Allow one pod (only) to be disrupted on a healthy cluster.
```

Comment:
Do we want to handle single ingest case too?

Reply:
Good point. Easy to add so let's do it!